dace/codegen/targets/fpga.py

```diff
@@ -1411,10 +1485,50 @@ def copy_memory(self, sdfg, dfg, state_id, src_node, dst_node, edge,
         dst_schedule = None if dst_parent is None else dst_parent.map.schedule
         state_dfg = sdfg.nodes()[state_id]
 
+        #Check if this is a copy memlet using at least one HBM array
+        do_default_copy = True
+        if (isinstance(src_node, dace.sdfg.nodes.AccessNode)
+                and isinstance(dst_node, dace.sdfg.nodes.AccessNode)):
+            src_array = src_node.desc(sdfg)
+            dst_array = dst_node.desc(sdfg)
+            src_is_hbm = utils.is_hbm_array(src_array)
+            dst_is_hbm = utils.is_hbm_array(dst_array)
+            if src_is_hbm or dst_is_hbm:
+                do_default_copy = False
+                modedge = copy.deepcopy(edge)
+                mem: memlet.Memlet = modedge.data
+                if mem.src_subset is None:
+                    mem.src_subset = subsets.Range.from_array(src_array)
+                if mem.dst_subset is None:
+                    mem.dst_subset = subsets.Range.from_array(dst_array)
+                if src_is_hbm:
+                    bankbeg, bankend = utils.get_multibank_ranges_from_subset(
+                        mem.src_subset, sdfg)
+                if dst_is_hbm:
+                    bankbeg, bankend = utils.get_multibank_ranges_from_subset(
+                        mem.dst_subset, sdfg)
+                num_accessed_banks = bankend - bankbeg
+                oldmem = copy.deepcopy(mem)
+                for i in range(num_accessed_banks):
+                    src_index = oldmem.src_subset[0][0] + i
+                    dst_index = oldmem.dst_subset[0][0] + i
+                    #Support for ignoring the distributed index if it's not required, e.g. on the host
+                    if src_is_hbm or num_accessed_banks > 1:
+                        mem.src_subset = utils.modify_distributed_subset(
+                            mem.src_subset, src_index)
+                    if dst_is_hbm or num_accessed_banks > 1:
+                        mem.dst_subset = utils.modify_distributed_subset(
+                            mem.dst_subset, dst_index)
+                    self._emit_copy(sdfg, state_id, src_node, src_storage,
+                                    dst_node, dst_storage, dst_schedule,
+                                    modedge, state_dfg, function_stream,
+                                    callsite_stream)
+
         # Emit actual copy
-        self._emit_copy(sdfg, state_id, src_node, src_storage, dst_node,
-                        dst_storage, dst_schedule, edge, state_dfg,
-                        function_stream, callsite_stream)
+        if do_default_copy:
+            self._emit_copy(sdfg, state_id, src_node, src_storage, dst_node,
+                            dst_storage, dst_schedule, edge, state_dfg,
+                            function_stream, callsite_stream)
```

Comment:
Did it now via a list, but not 100% sure that this is nicer. I don't want to somehow handle normal copies in the section where HBM copies are handled at the moment, because this would require handling quite a few special conditions in that part, which will make it harder to read

Reply:
Ah I see, I didn't consider that the HBM one is in a loop. Anyway, I'm okay with either solution then 🤷